### PR TITLE
Add support for Python 3.12; drop support for Python 3.8

### PR DESCRIPTION
**Pull request by fmhoeger**
Support Python versions 3.9, 3.10, 3.11, and 3.12.

* Add support for Python 3.12
* Remove support for Python 3.8
* Remove pinning of packages due to Python 3.8 incompatibility, see https://github.com/Dallinger/Dallinger/pull/5545
* Run `Full tox` on both Python 3.11 and 3.12

### CODE REVIEW

**Comment by fmhoeger** [comment]
@pmcharrison As you already agreed on dropping Python 3.8 here: https://github.com/Dallinger/Dallinger/issues/5205#issuecomment-1908931431
we could use this PR to additionally support Python 3.12 and close https://github.com/Dallinger/Dallinger/pull/5545

**Comment by pmcharrison** [comment]
Yes, I would be happy with that

**Comment by fmhoeger** [comment]
Great, so i am waiting for your approval then.

File: .github/workflows/ci.yml
Comment (pmcharrison) [inline .github/workflows/ci.yml:116]: I don't really see the logic of this and the below `python-version` bits. I think the logic is that we should only run certain 'full' tests on the latest version of Python; we shouldn't run them on every version above and including 3.11. I think we should just be replacing `3.11` with `3.12` and leaving everything else the same, no?

Comment (fmhoeger) [reply .github/workflows/ci.yml:116]: OK, I thought it might be more secure to run the full tests on the two latest versions as there is not much additional cost and we can be sure that people still on the recent latest version of Python (in this case 3.11) are assured the are no issues with code changes that had to be made for the Python version upgrade.

Comment (fmhoeger) [reply .github/workflows/ci.yml:116]: See for example [this commit](https://github.com/Dallinger/Dallinger/pull/6013/commits/5ceff4150986ac1e22233a1ea6efc7ab1b61e035) which required a code change to cope with breaking API change: 

The best would probably be to run the full tests on all versions. But i didn't want to go that far.

**Comment by pmcharrison** [comment]
I see, maybe just add a comment to the file to explain your logic then so that future people aren't confused?
________________________________

**Review by pmcharrison — APPROVED** [review]

**Pull request merged**
